lib/Service/LoginService.php

```diff
@@ -206,7 +209,12 @@ public function completeLogin($user, $password, $userSession, $request)
 
         // Update the user's last login timestamp, since the conditions above tend to cause the
         // completeLogin() call above to skip doing so.
-        $user->updateLastLoginTimestamp();
+        $firstTimeLogin = $user->updateLastLoginTimestamp();
+
+        // Same for the firstLogin event
+        if ($firstTimeLogin) {
+            \OC::$server->get(IEventDispatcher::class)->dispatch(IUser::class.'::firstLogin', new GenericEvent($user));
```

Comment:
How about calling `Session::prepareUserLogin` instead? This seems like a very internal event.